// instantsearch.js/cjs/widgets/chat-trigger/chat-trigger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return chatTrigger;
    }
});
var _interop_require_default = require("@swc/helpers/_/_interop_require_default");
var _object_spread = require("@swc/helpers/_/_object_spread");
var _object_spread_props = require("@swc/helpers/_/_object_spread_props");
var _instantsearchuicomponents = require("instantsearch-ui-components");
var _preact = require("preact");
var _Template = /*#__PURE__*/ _interop_require_default._(require("../../components/Template/Template"));
var _connectChatTrigger = /*#__PURE__*/ _interop_require_default._(require("../../connectors/chat/connectChatTrigger"));
var _templating = require("../../lib/templating");
var _utils = require("../../lib/utils");
var withUsage = (0, _utils.createDocumentationMessageGenerator)({
    name: 'chatTrigger'
});
var ChatToggleButton = (0, _instantsearchuicomponents.createChatToggleButtonComponent)({
    createElement: _preact.h,
    Fragment: _preact.Fragment
});
function chatTrigger(widgetParams) {
    var _userCssClasses_button;
    var _ref = widgetParams || {}, container = _ref.container, tmp = _ref.templates, userTemplates = tmp === void 0 ? {} : tmp, tmp1 = _ref.cssClasses, userCssClasses = tmp1 === void 0 ? {} : tmp1, _ref_floating = _ref.floating, floating = _ref_floating === void 0 ? true : _ref_floating;
    var rootClassName = floating ? [
        'ais-ChatToggleButton--floating'
    ].concat((_userCssClasses_button = userCssClasses.button) !== null && _userCssClasses_button !== void 0 ? _userCssClasses_button : []) : userCssClasses.button;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = (0, _utils.getContainerNode)(container);
    var templates = _object_spread._({}, userTemplates);
    function renderTrigger(renderState, _isFirstRender) {
        var open = renderState.open, toggleOpen = renderState.toggleOpen, instantSearchInstance = renderState.instantSearchInstance;
        // Resolve template props at render time so user-provided helpers and
        // compile options from `instantSearchInstance.templatesConfig` apply.
        var templateProps = (0, _templating.prepareTemplateProps)({
            defaultTemplates: {},
            templatesConfig: instantSearchInstance.templatesConfig,
            templates: templates
        });
        var LayoutComponent = templates.layout ? function(props) {
            return /*#__PURE__*/ (0, _preact.h)(_Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
                templateKey: "layout",
                rootTagName: "fragment",
                data: props
            }));
        } : undefined;
        var iconComponent = templates.icon ? function(param) {
            var isOpen = param.isOpen;
            return /*#__PURE__*/ (0, _preact.h)(_Template.default, _object_spread_props._(_object_spread._({}, templateProps), {
                templateKey: "icon",
                rootTagName: "span",
                data: {
                    isOpen: isOpen
                }
            }));
        } : undefined;
        if (LayoutComponent) {
            (0, _preact.render)(/*#__PURE__*/ (0, _preact.h)(LayoutComponent, {
                open: open,
                onClick: toggleOpen,
                toggleIconComponent: iconComponent
            }), containerNode);
        } else {
            (0, _preact.render)(/*#__PURE__*/ (0, _preact.h)(ChatToggleButton, {
                open: open,
                onClick: toggleOpen,
                toggleIconComponent: iconComponent,
                classNames: {
                    root: rootClassName
                }
            }), containerNode);
        }
    }
    var makeWidget = (0, _connectChatTrigger.default)(renderTrigger, function() {
        return (0, _preact.render)(null, containerNode);
    });
    return _object_spread_props._(_object_spread._({}, makeWidget({})), {
        $$widgetType: 'ais.chatTrigger'
    });
}
