// instantsearch.js/es/templates/chat-layout/chat-inline-layout.js
import { _ as _$2 } from '@swc/helpers/esm/_object_spread.js';
import { _ as _$1 } from '@swc/helpers/esm/_object_spread_props.js';
import { _ } from '@swc/helpers/esm/_object_without_properties.js';
import { createChatInlineLayoutComponent } from 'instantsearch-ui-components';
import { h, Fragment } from 'preact';

var ChatInlineLayout = createChatInlineLayoutComponent({
    createElement: h,
    Fragment: Fragment
});
function chatInlineLayout() {
    function ChatInlineLayoutTemplate(props) {
        var templates = props.templates, rest = _(props, [
            "templates"
        ]);
        return /*#__PURE__*/ h(ChatInlineLayout, _$1(_$2({}, rest), {
            headerComponent: templates.header(),
            messagesComponent: templates.messages(),
            promptComponent: templates.prompt()
        }));
    }
    // Marker used by `chat` to auto-exempt inline layouts from the entry-point
    // validation, since inline chats are always visible (no trigger needed).
    ChatInlineLayoutTemplate.$$inlineLayout = true;
    return ChatInlineLayoutTemplate;
}

export { chatInlineLayout };
