// instantsearch-ui-components/dist/es/components/autocomplete/AutocompleteSearch.js
import { _ } from '@swc/helpers/esm/_object_spread.js';
import { cx } from '../../lib/cx.js';
import { BackIcon, SubmitIcon, LoadingIcon, ClearIcon, AiModeIcon } from './icons.js';

function createAutocompleteSearchComponent(param) {
    var createElement = param.createElement;
    return function AutocompleteSearch(userProps) {
        var inputProps = userProps.inputProps, onClear = userProps.onClear, query = userProps.query, isSearchStalled = userProps.isSearchStalled, onCancel = userProps.onCancel, isDetached = userProps.isDetached, submitTitle = userProps.submitTitle, onAiModeClick = userProps.onAiModeClick, _userProps_aiModeButtonDisabled = userProps.aiModeButtonDisabled, aiModeButtonDisabled = _userProps_aiModeButtonDisabled === void 0 ? false : _userProps_aiModeButtonDisabled, _userProps_classNames = userProps.classNames, classNames = _userProps_classNames === void 0 ? {} : _userProps_classNames;
        var isBackButton = Boolean(isDetached && onCancel);
        var resolvedCancelTitle = submitTitle !== null && submitTitle !== void 0 ? submitTitle : 'Close';
        var inputRef = inputProps.ref;
        return /*#__PURE__*/ createElement("form", {
            className: cx('ais-AutocompleteForm', classNames.form),
            action: "",
            noValidate: true,
            role: "search",
            onSubmit: function onSubmit(e) {
                e.preventDefault();
            },
            onReset: function onReset() {
                var _inputRef_current;
                return (_inputRef_current = inputRef.current) === null || _inputRef_current === void 0 ? void 0 : _inputRef_current.focus();
            }
        }, /*#__PURE__*/ createElement("div", {
            className: cx('ais-AutocompleteInputWrapperPrefix', classNames.inputWrapperPrefix)
        }, isBackButton && /*#__PURE__*/ createElement("button", {
            className: cx('ais-AutocompleteBackButton', classNames.backButton),
            type: "button",
            title: resolvedCancelTitle,
            onClick: onCancel,
            hidden: isSearchStalled
        }, /*#__PURE__*/ createElement(BackIcon, {
            createElement: createElement,
            className: classNames.backButtonIcon
        })), /*#__PURE__*/ createElement("label", {
            className: cx('ais-AutocompleteLabel', classNames.label),
            "aria-label": "Submit",
            htmlFor: inputProps.id,
            id: "".concat(inputProps.id, "-label"),
            hidden: isBackButton || undefined
        }, /*#__PURE__*/ createElement("button", {
            className: cx('ais-AutocompleteSubmitButton', classNames.submitButton),
            type: "submit",
            title: "Submit",
            hidden: isSearchStalled
        }, /*#__PURE__*/ createElement(SubmitIcon, {
            createElement: createElement,
            className: classNames.submitButtonIcon
        }))), /*#__PURE__*/ createElement("div", {
            className: cx('ais-AutocompleteLoadingIndicator', classNames.loadingIndicator),
            hidden: !isSearchStalled
        }, /*#__PURE__*/ createElement(LoadingIcon, {
            createElement: createElement,
            isSearchStalled: isSearchStalled,
            className: classNames.loadingIndicatorIcon
        }))), /*#__PURE__*/ createElement("div", {
            className: cx('ais-AutocompleteInputWrapper', classNames.inputWrapper)
        }, /*#__PURE__*/ createElement("input", _({
            className: cx('ais-AutocompleteInput', classNames.input),
            "aria-autocomplete": "both",
            "aria-labelledby": "".concat(inputProps.id, "-label"),
            autoComplete: "off",
            autoCorrect: "off",
            autoCapitalize: "off",
            enterKeyHint: "search",
            spellCheck: "false",
            maxLength: 512,
            type: "search",
            value: query
        }, inputProps))), /*#__PURE__*/ createElement("div", {
            className: cx('ais-AutocompleteInputWrapperSuffix', classNames.inputWrapperSuffix)
        }, /*#__PURE__*/ createElement("button", {
            className: cx('ais-AutocompleteClearButton', classNames.resetButton),
            type: "reset",
            title: "Clear",
            hidden: query.length === 0 || isSearchStalled,
            onClick: onClear
        }, /*#__PURE__*/ createElement(ClearIcon, {
            createElement: createElement,
            className: classNames.resetButtonIcon
        })), onAiModeClick && /*#__PURE__*/ createElement("button", {
            className: cx('ais-AiModeButton', classNames.aiModeButton),
            type: "button",
            title: "AI Mode",
            disabled: aiModeButtonDisabled,
            onClick: function onClick(e) {
                e.preventDefault();
                onAiModeClick();
            }
        }, /*#__PURE__*/ createElement(AiModeIcon, {
            createElement: createElement,
            className: classNames.aiModeButtonIcon
        }), /*#__PURE__*/ createElement("span", {
            className: cx('ais-AiModeButton-label', classNames.aiModeButtonLabel)
        }, "AI Mode"))));
    };
}

export { createAutocompleteSearchComponent };
