// instantsearch.js/es/widgets/chat-trigger/chat-trigger.js
import { _ } from '@swc/helpers/esm/_object_spread.js';
import { _ as _$1 } from '@swc/helpers/esm/_object_spread_props.js';
import { createChatToggleButtonComponent } from 'instantsearch-ui-components';
import { h, Fragment, render } from 'preact';
import Template from '../../components/Template/Template.js';
import connectChatTrigger from '../../connectors/chat/connectChatTrigger.js';
import { createDocumentationMessageGenerator } from '../../lib/utils/documentation.js';
import { getContainerNode } from '../../lib/utils/getContainerNode.js';
import { prepareTemplateProps } from '../../lib/templating/prepareTemplateProps.js';

var withUsage = createDocumentationMessageGenerator({
    name: 'chatTrigger'
});
var ChatToggleButton = createChatToggleButtonComponent({
    createElement: h,
    Fragment: Fragment
});
function chatTrigger(widgetParams) {
    var _userCssClasses_button;
    var _ref = widgetParams || {}, container = _ref.container, tmp = _ref.templates, userTemplates = tmp === void 0 ? {} : tmp, tmp1 = _ref.cssClasses, userCssClasses = tmp1 === void 0 ? {} : tmp1, _ref_floating = _ref.floating, floating = _ref_floating === void 0 ? true : _ref_floating;
    var rootClassName = floating ? [
        'ais-ChatToggleButton--floating'
    ].concat((_userCssClasses_button = userCssClasses.button) !== null && _userCssClasses_button !== void 0 ? _userCssClasses_button : []) : userCssClasses.button;
    if (!container) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    var containerNode = getContainerNode(container);
    var templates = _({}, userTemplates);
    function renderTrigger(renderState, _isFirstRender) {
        var open = renderState.open, toggleOpen = renderState.toggleOpen, instantSearchInstance = renderState.instantSearchInstance;
        // Resolve template props at render time so user-provided helpers and
        // compile options from `instantSearchInstance.templatesConfig` apply.
        var templateProps = prepareTemplateProps({
            defaultTemplates: {},
            templatesConfig: instantSearchInstance.templatesConfig,
            templates: templates
        });
        var LayoutComponent = templates.layout ? function(props) {
            return /*#__PURE__*/ h(Template, _$1(_({}, templateProps), {
                templateKey: "layout",
                rootTagName: "fragment",
                data: props
            }));
        } : undefined;
        var iconComponent = templates.icon ? function(param) {
            var isOpen = param.isOpen;
            return /*#__PURE__*/ h(Template, _$1(_({}, templateProps), {
                templateKey: "icon",
                rootTagName: "span",
                data: {
                    isOpen: isOpen
                }
            }));
        } : undefined;
        if (LayoutComponent) {
            render(/*#__PURE__*/ h(LayoutComponent, {
                open: open,
                onClick: toggleOpen,
                toggleIconComponent: iconComponent
            }), containerNode);
        } else {
            render(/*#__PURE__*/ h(ChatToggleButton, {
                open: open,
                onClick: toggleOpen,
                toggleIconComponent: iconComponent,
                classNames: {
                    root: rootClassName
                }
            }), containerNode);
        }
    }
    var makeWidget = connectChatTrigger(renderTrigger, function() {
        return render(null, containerNode);
    });
    return _$1(_({}, makeWidget({})), {
        $$widgetType: 'ais.chatTrigger'
    });
}

export { chatTrigger as default };
