// instantsearch.js/es/lib/ai-lite/utils.js
import { _ as _$2 } from '@swc/helpers/esm/_call_super.js';
import { _ as _$1 } from '@swc/helpers/esm/_class_call_check.js';
import { _ as _$4 } from '@swc/helpers/esm/_create_class.js';
import { _ as _$6 } from '@swc/helpers/esm/_define_property.js';
import { _ } from '@swc/helpers/esm/_inherits.js';
import { _ as _$5 } from '@swc/helpers/esm/_type_of.js';
import { _ as _$3 } from '@swc/helpers/esm/_wrap_native_super.js';

function generateId() {
    return Math.random().toString(36).substring(2, 9);
}
function isToolOrDynamicToolUIPart(part) {
    if ((typeof part === "undefined" ? "undefined" : _$5(part)) !== 'object' || part === null) return false;
    var p = part;
    return typeof p.type === 'string' && (p.type.startsWith('tool-') || p.type === 'dynamic-tool');
}
function lastAssistantMessageIsCompleteWithToolCalls(param) {
    var messages = param.messages;
    if (messages.length === 0) return false;
    var lastMessage = messages[messages.length - 1];
    if (lastMessage.role !== 'assistant') return false;
    if (!lastMessage.parts || lastMessage.parts.length === 0) return false;
    var toolParts = lastMessage.parts.filter(isToolOrDynamicToolUIPart);
    if (toolParts.length === 0) return false;
    return toolParts.every(function(part) {
        return part.state === 'output-available' || part.state === 'output-error';
    });
}
var SerialJobExecutor = /*#__PURE__*/ function() {
    function SerialJobExecutor() {
        _$1(this, SerialJobExecutor);
        _$6(this, "queue", []);
        _$6(this, "isRunning", false);
    }
    _$4(SerialJobExecutor, [
        {
            key: "run",
            value: function run(job) {
                var _this = this;
                return new Promise(function(resolve, reject) {
                    _this.queue.push(function() {
                        return job().then(function(result) {
                            resolve(result);
                        }, function(error) {
                            reject(error);
                        });
                    });
                    _this.processQueue();
                });
            }
        },
        {
            key: "processQueue",
            value: function processQueue() {
                var _this = this;
                if (this.isRunning) return;
                this.isRunning = true;
                var processNext = function processNext1() {
                    if (_this.queue.length === 0) {
                        _this.isRunning = false;
                        return;
                    }
                    var job = _this.queue.shift();
                    if (job) {
                        job().then(processNext, processNext);
                    }
                };
                processNext();
            }
        }
    ]);
    return SerialJobExecutor;
}();
function resolveValue(value) {
    if (value === undefined) return Promise.resolve(undefined);
    if (typeof value === 'function') {
        return Promise.resolve(value());
    }
    return Promise.resolve(value);
}
/**
 * Error raised when the upstream stream emits a `data-guardrail-violation`
 * chunk. The `message` carries the service-provided `fallbackResponse`, which
 * is intentionally surfaced verbatim through the chat error UI (unlike
 * generic cost-control / 4xx errors, where the raw API message is hidden
 * behind a friendly default).
 *
 * Detection across package boundaries should rely on `error.name` rather than
 * `instanceof` to avoid issues with mixed module copies in bundled apps.
 */ var GuardrailViolationError = /*#__PURE__*/ function(Error1) {
    _(GuardrailViolationError, Error1);
    function GuardrailViolationError(message) {
        _$1(this, GuardrailViolationError);
        var _this;
        _this = _$2(this, GuardrailViolationError, [
            message
        ]);
        _this.name = 'GuardrailViolationError';
        return _this;
    }
    return GuardrailViolationError;
}(_$3(Error));
/**
 * Reads a non-empty `message` field off a JSON-serialized error envelope.
 *
 * Both transports backing `AbstractChat` (stream `error` chunks and HTTP error
 * responses) serialize errors as `{"message": "...", ...}` — the same shape as
 * the shared `ErrorResponse` on the API side. Returns the trimmed message if
 * the input is such a JSON object, otherwise `undefined`.
 */ function tryParseErrorMessage(text) {
    try {
        var parsed = JSON.parse(text);
        if (parsed && (typeof parsed === "undefined" ? "undefined" : _$5(parsed)) === 'object' && !Array.isArray(parsed) && typeof parsed.message === 'string') {
            var message = parsed.message.trim();
            if (message) {
                return message;
            }
        }
    } catch (unused) {
    // Not JSON — caller falls back to its own default.
    }
    return undefined;
}

export { GuardrailViolationError, SerialJobExecutor, generateId, lastAssistantMessageIsCompleteWithToolCalls, resolveValue, tryParseErrorMessage };
