// @meilisearch/instant-meilisearch/dist/index.js
import { Meilisearch as e } from "meilisearch";
//#region src/utils/object.ts
function t(e) {
	return typeof e == "object" && !Array.isArray(e) && e !== null;
}
//#endregion
//#region src/client/config/index.ts
function n(e) {
	return {
		placeholderSearch: !0,
		keepZeroFacets: !1,
		clientAgents: [],
		finitePagination: !1,
		...e
	};
}
function r(e) {
	if (typeof e == "function") {
		let t = e();
		if (typeof t != "string") throw TypeError("Provided apiKey function (2nd parameter) did not return a string, expected string");
		return t;
	}
	return e;
}
function i(e, n, r) {
	let { requestInit: i, httpClient: a } = r;
	if (typeof e != "string") throw TypeError("Provided hostUrl value (1st parameter) is not a string, expected string");
	if (typeof n != "string" && typeof n != "function") throw TypeError("Provided apiKey value (2nd parameter) is not a string or a function, expected string or function");
	if (i !== void 0 && !t(i)) throw TypeError("Provided requestInit should be an object");
	if (a && typeof a != "function") throw TypeError("Provided custom httpClient should be a function");
}
//#endregion
//#region src/adapter/search-request-adapter/search-resolver.ts
function a(e, t) {
	return { multiSearch: async function(n, r) {
		let i = t.formatKey([n]), a = t.getEntry(i);
		if (a) return a;
		let o = (await e.multiSearch({ queries: n })).results.map((e, t) => ({
			...e,
			pagination: r[t] || {}
		}));
		return t.setEntry(i, o), o;
	} };
}
//#endregion
//#region src/adapter/search-request-adapter/geo-rules-adapter.ts
function o({ insidePolygon: e, insideBoundingBox: t, aroundLatLng: n, aroundRadius: r, minimumAroundRadius: i }) {
	let a, o, s;
	if (Array.isArray(e) && e.length >= 3) {
		let t = [], n = e.map((e) => {
			if (!Array.isArray(e) || e.length < 2) return t.push(e), null;
			let n = Number.parseFloat(String(e[0])), r = Number.parseFloat(String(e[1]));
			return Number.isNaN(n) || Number.isNaN(r) ? (t.push(e), null) : `[${n}, ${r}]`;
		}).filter((e) => e !== null);
		if (t.length > 0 && console.warn("instant-meilisearch: insidePolygon contains invalid coordinate pairs that were ignored:", t), n.length >= 3) return s = `_geoPolygon(${n.join(", ")})`, s;
	}
	if (n) {
		let [e, t] = n.split(",").map((e) => Number.parseFloat(e).toFixed(5));
		a = [e, t];
	}
	if ((r != null || i != null) && (r === "all" ? console.warn("instant-meilisearch is not compatible with the `all` value on the aroundRadius parameter") : o = r ?? i), t && typeof t == "string") {
		let [e, n, r, i] = t.split(",").map((e) => parseFloat(e));
		s = `_geoBoundingBox([${e}, ${n}], [${r}, ${i}])`;
	} else if (a != null && o != null) {
		let [e, t] = a;
		s = `_geoRadius(${e}, ${t}, ${o})`;
	}
	return s;
}
//#endregion
//#region src/adapter/search-request-adapter/filter-adapter.ts
var s = /([\\"])/g;
function c(e) {
	return e.replace(s, "\\$1");
}
function l(e) {
	let t = c(e), n = t.indexOf(":");
	return `"${t.slice(0, n)}"="${t.slice(n + 1)}"`;
}
function u(e) {
	let [t, n, r] = (() => {
		let t = e.match(/^([^<!>:=]*)([<!>:=]+)(.*)$/);
		if (t) {
			let [e, n, r] = t.slice(1), [i, a] = (n.match(/^([<!>]?=|<|>|:){1}(.*)/) || ["", ""]).slice(1);
			return [
				e,
				i,
				a + r
			];
		}
		return [
			e,
			"",
			""
		];
	})();
	return `"${c(t).trim()}"${n === ":" ? " " : n}${r.trim()}`;
}
function d(e, t) {
	return typeof t == "string" ? e(t) : t.map((t) => typeof t == "string" ? e(t) : t.map((t) => e(t)));
}
function f(e) {
	return typeof e == "string" ? [e] : e;
}
function p(e, t, n) {
	let r = f(t), i = f(n), a = [];
	return e !== void 0 && a.push(e), r !== void 0 && a.push(...r), i !== void 0 && a.push(...i), a;
}
function m(e, t, n) {
	return p(e, t === void 0 ? t : d(u, t), n === void 0 ? n : d(l, n));
}
//#endregion
//#region src/adapter/search-request-adapter/search-params-adapter.ts
function h(e, t, n) {
	return !(!n && !t && (!e || e.length === 0));
}
function g(e, t) {
	let { page: n, hitsPerPage: r } = e;
	return t ? {
		limit: r + 1,
		offset: n * r
	} : {
		limit: 0,
		offset: 0
	};
}
function _(e, t) {
	let { page: n, hitsPerPage: r } = e;
	return t ? {
		hitsPerPage: r,
		page: n + 1
	} : {
		hitsPerPage: 0,
		page: n + 1
	};
}
function v(e) {
	let { query: t, indexUid: n, facets: r, attributesToSnippet: i, snippetEllipsisText: a, filters: s, numericFilters: c, facetFilters: l, attributesToRetrieve: u, attributesToHighlight: d, highlightPreTag: f, highlightPostTag: p, placeholderSearch: v, pagination: y, sort: b, restrictSearchableAttributes: x, meiliSearchParams: S } = e, C = { indexUid: n }, w = m(s, c, l);
	return {
		getParams() {
			return C;
		},
		addQuery() {
			C.q = t;
		},
		addFacets() {
			let e = S?.indexesOverrides?.[n]?.facets ?? S?.facets ?? r;
			e !== void 0 && (C.facets = typeof e == "string" ? [e] : e);
		},
		addAttributesToCrop() {
			let e = S?.indexesOverrides?.[n]?.attributesToCrop ?? S?.attributesToCrop ?? i;
			e !== void 0 && (C.attributesToCrop = e);
		},
		addCropLength() {
			let e = S?.indexesOverrides?.[n]?.cropLength ?? S?.cropLength;
			e !== void 0 && (C.cropLength = e);
		},
		addCropMarker() {
			let e = S?.indexesOverrides?.[n]?.cropMarker ?? S?.cropMarker ?? a;
			e !== void 0 && (C.cropMarker = e);
		},
		addFilters() {
			let e = S?.indexesOverrides?.[n]?.filter ?? S?.filter;
			e === void 0 ? w.length && (C.filter = w) : C.filter = e;
		},
		addAttributesToRetrieve() {
			let e = S?.indexesOverrides?.[n]?.attributesToRetrieve ?? S?.attributesToRetrieve ?? u;
			e !== void 0 && (C.attributesToRetrieve = e);
		},
		addAttributesToHighlight() {
			C.attributesToHighlight = S?.indexesOverrides?.[n]?.attributesToHighlight ?? S?.attributesToHighlight ?? d ?? ["*"];
		},
		addPreTag() {
			C.highlightPreTag = S?.indexesOverrides?.[n]?.highlightPreTag ?? S?.highlightPreTag ?? f ?? "__ais-highlight__";
		},
		addPostTag() {
			C.highlightPostTag = S?.indexesOverrides?.[n]?.highlightPostTag ?? S?.highlightPostTag ?? p ?? "__/ais-highlight__";
		},
		addPagination() {
			let e = h(w, t, v);
			if (y.finite) {
				let { hitsPerPage: t, page: n } = _(y, e);
				C.hitsPerPage = t, C.page = n;
			} else {
				let { limit: t, offset: n } = g(y, e);
				C.limit = t, C.offset = n;
			}
		},
		addSort() {
			let e = S?.indexesOverrides?.[n]?.sort ?? S?.sort ?? b;
			e && (!Array.isArray(e) || e.length) && (C.sort = Array.isArray(e) ? e : [e]);
		},
		addGeoSearchFilter() {
			let { insidePolygon: t, insideBoundingBox: n, aroundLatLng: r, aroundRadius: i, minimumAroundRadius: a } = e, s = o({
				insidePolygon: t,
				insideBoundingBox: n,
				aroundLatLng: r,
				aroundRadius: i,
				minimumAroundRadius: a
			});
			s !== void 0 && (Array.isArray(C.filter) ? C.filter.unshift(s) : C.filter = [s]);
		},
		addShowMatchesPosition() {
			let e = S?.indexesOverrides?.[n]?.showMatchesPosition ?? S?.showMatchesPosition;
			e !== void 0 && (C.showMatchesPosition = e);
		},
		addMatchingStrategy() {
			let e = S?.indexesOverrides?.[n]?.matchingStrategy ?? S?.matchingStrategy;
			e !== void 0 && (C.matchingStrategy = e);
		},
		addShowRankingScore() {
			let e = S?.indexesOverrides?.[n]?.showRankingScore ?? S?.showRankingScore;
			e !== void 0 && (C.showRankingScore = e);
		},
		addAttributesToSearchOn() {
			let e = S?.indexesOverrides?.[n]?.attributesToSearchOn ?? S?.attributesToSearchOn ?? x;
			e !== void 0 && (C.attributesToSearchOn = e);
		},
		addHybridSearch() {
			let e = S?.indexesOverrides?.[n]?.hybrid ?? S?.hybrid;
			e !== void 0 && (C.hybrid = e);
		},
		addVector() {
			let e = S?.indexesOverrides?.[n]?.vector ?? S?.vector;
			e !== void 0 && (C.vector = e);
		},
		addDistinct() {
			let e = S?.indexesOverrides?.[n]?.distinct ?? S?.distinct;
			e !== void 0 && (C.distinct = e);
		},
		addRankingScoreThreshold() {
			let e = S?.indexesOverrides?.[n]?.rankingScoreThreshold ?? S?.rankingScoreThreshold;
			e !== void 0 && (C.rankingScoreThreshold = e);
		}
	};
}
function y(e) {
	let t = v(e);
	return t.addQuery(), t.addFacets(), t.addAttributesToCrop(), t.addCropLength(), t.addCropMarker(), t.addFilters(), t.addAttributesToRetrieve(), t.addAttributesToHighlight(), t.addPreTag(), t.addPostTag(), t.addPagination(), t.addSort(), t.addGeoSearchFilter(), t.addShowMatchesPosition(), t.addMatchingStrategy(), t.addShowRankingScore(), t.addAttributesToSearchOn(), t.addHybridSearch(), t.addVector(), t.addDistinct(), t.addRankingScoreThreshold(), t.getParams();
}
//#endregion
//#region src/utils/array.ts
function b(e) {
	let t = [];
	return (n) => t.includes(n[e]) ? !1 : (t.push(n[e]), !0);
}
//#endregion
//#region src/utils/string.ts
function x(e) {
	return e.reduce((e, t) => e + JSON.stringify(t), "");
}
//#endregion
//#region src/adapter/search-response-adapter/format-adapter/format-adapter.ts
function S(e) {
	return typeof e == "string" ? e : JSON.stringify(e === void 0 ? null : e);
}
function C(e) {
	return Array.isArray(e) ? e.map((e) => C(e)) : t(e) ? Object.keys(e).reduce((t, n) => (t[n] = C(e[n]), t), {}) : { value: S(e) };
}
function w(e) {
	if (!e) return {};
	let t = C(e);
	return {
		_highlightResult: t,
		_snippetResult: t
	};
}
//#endregion
//#region src/adapter/search-response-adapter/geo-reponse-adapter.ts
function T(e) {
	for (let t = 0; t < e.length; t++) {
		let n = `${t + Math.random() * 1e6}`;
		e[t]._geo && (e[t]._geoloc = e[t]._geo, e[t].objectID || (e[t].objectID = n)), e[t]._formatted?._geo && (e[t]._formatted._geoloc = e[t]._formatted._geo, e[t]._formatted.objectID || (e[t]._formatted.objectID = n));
	}
	return e;
}
//#endregion
//#region src/adapter/search-response-adapter/hits-adapter.ts
function E(e, t) {
	let { hits: n } = e, { hitsPerPage: r, page: i } = e.pagination, { finitePagination: a, primaryKey: o } = t;
	!a && n.length > r && n.splice(n.length - 1, 1);
	let s = n.map((e, n) => {
		if (Object.keys(e).length > 0) {
			let { _formatted: a, _matchesPosition: s, ...c } = e, l = Object.assign(c, w(a));
			return t?.meiliSearchParams?.showMatchesPosition && (l._matchesPosition = s), o && (l.objectID = e[o]), l.__position = i * r + n + 1, l;
		}
		return e;
	});
	return s = T(s), s;
}
//#endregion
//#region src/adapter/search-response-adapter/total-hits-adapter.ts
function D(e) {
	let { hitsPerPage: t = 0, totalPages: n = 0, estimatedTotalHits: r, totalHits: i } = e;
	return r ?? i ?? t * n;
}
//#endregion
//#region src/adapter/search-response-adapter/pagination-adapter.ts
function O(e, t) {
	if (e.totalPages != null) return e.totalPages;
	if (t === 0) return 0;
	let { limit: n = 20, offset: r = 0, hits: i } = e, a = +(i.length >= n);
	return r / t + 1 + a;
}
function k(e, t) {
	let { hitsPerPage: n, page: r } = t;
	return {
		page: r,
		nbPages: O(e, n),
		hitsPerPage: n
	};
}
//#endregion
//#region src/adapter/search-response-adapter/facet-distribution-adapter.ts
function A(e) {
	return e ? typeof e == "string" ? [e] : e : [];
}
function j(e, t, n) {
	let r = A(e), i = {};
	for (let e of r) for (let r in t[e]) i[e] || (i[e] = n[e] || {}), i[e][r] ? i[e][r] = n[e][r] : i[e][r] = 0;
	return i;
}
function M(e, t, n, r) {
	return e ? (r ||= {}, j(t, n, r)) : r;
}
//#endregion
//#region src/adapter/search-response-adapter/adapt-facet-stats.ts
function N(e) {
	return Object.keys(e).reduce((t, n) => (t[n] = {
		...e[n],
		avg: 0,
		sum: 0
	}, t), {});
}
//#endregion
//#region src/adapter/search-response-adapter/search-response-adapter.ts
function P(e, t, n) {
	return { results: e.map((e) => F(e, t[e.indexUid], n)) };
}
function F(e, t, n) {
	let { processingTimeMs: r, query: i, indexUid: a, facetDistribution: o = {}, facetStats: s = {}, metadata: c } = e, l = Object.keys(o), { hitsPerPage: u, page: d, nbPages: f } = k(e, e.pagination), p = E(e, n), m = D(e), h = {
		index: a,
		hitsPerPage: u,
		page: d,
		facets: M(n.keepZeroFacets, l, t, o),
		nbPages: f,
		nbHits: m,
		processingTimeMS: r,
		query: i,
		hits: p,
		params: "",
		exhaustiveNbHits: !1,
		facets_stats: N(s)
	};
	return c && (h._meilisearch = { metadata: c }), h;
}
//#endregion
//#region src/contexts/sort-context.ts
function I(e) {
	if (!e) return [];
	let t = /[^:]+:(?:asc|desc)/g, n = [], r;
	for (; (r = t.exec(e)) !== null;) n.push(r[0]);
	return n.map((e) => e.replace(/^,+|,+$/, ""));
}
//#endregion
//#region src/contexts/pagination-context.ts
function L(e, t, n) {
	return {
		hitsPerPage: t === void 0 ? 20 : t,
		page: n || 0,
		finite: !!e
	};
}
//#endregion
//#region src/contexts/search-context.ts
function R(e) {
	let t = e.indexOf(":");
	return t === -1 ? {
		indexUid: e,
		sortBy: ""
	} : {
		indexUid: e.substring(0, t),
		sortBy: e.substring(t + 1)
	};
}
function z(e, t) {
	let { query: n, indexName: r, params: i } = e, { indexUid: a, sortBy: o } = R(r), s = L(t.finitePagination, i?.hitsPerPage, i?.page);
	return {
		...t,
		query: n,
		...i,
		sort: I(o),
		indexUid: a,
		pagination: s,
		placeholderSearch: t.placeholderSearch !== !1,
		keepZeroFacets: !!t.keepZeroFacets
	};
}
function B(e, t) {
	let { indexUid: n, sortBy: r } = R(e.indexName), { params: i } = e, a = L(t.finitePagination, i?.hitsPerPage, i?.page);
	return {
		...t,
		...i,
		sort: I(r),
		indexUid: n,
		pagination: a,
		placeholderSearch: t.placeholderSearch !== !1,
		keepZeroFacets: !!t.keepZeroFacets
	};
}
//#endregion
//#region src/cache/search-cache.ts
function V(e = {}) {
	let t = e;
	return {
		getEntry: function(e) {
			if (t[e]) try {
				return JSON.parse(t[e]);
			} catch {
				return;
			}
		},
		formatKey: function(e) {
			return x(e);
		},
		setEntry: function(e, n) {
			t[e] = JSON.stringify(n);
		},
		clearCache: function() {
			t = {};
		}
	};
}
//#endregion
//#region src/cache/init-facets-distribution.ts
function H(e) {
	let t = v({
		...e,
		placeholderSearch: !0,
		query: ""
	});
	return t.addFacets(), t.addPagination(), t.getParams();
}
async function U(e, t, n) {
	let r = b("indexUid"), i = t.filter(r).filter(({ indexUid: e }) => !Object.keys(n).includes(e));
	if (i.length === 0) return n;
	let a = await e.multiSearch(i, []);
	for (let e of a) n[e.indexUid] = e.facetDistribution || {};
	return n;
}
function W(e, t) {
	for (let n of t) e[n.indexUid] = {
		...n.facetDistribution || {},
		...e[n.indexUid] || {}
	};
	return e;
}
var G = {
	name: "@meilisearch/instant-meilisearch",
	version: "0.31.2",
	description: "The search client to use Meilisearch with InstantSearch.",
	homepage: "https://github.com/meilisearch/meilisearch-js-plugins/tree/main/packages/instant-meilisearch",
	license: "MIT",
	bugs: { url: "https://github.com/meilisearch/meilisearch-js-plugins/issues" },
	repository: {
		type: "git",
		url: "https://github.com/meilisearch/meilisearch-js-plugins.git"
	},
	author: {
		name: "Meilisearch",
		url: "https://meilisearch.com"
	},
	keywords: [
		"search",
		"instantsearch",
		"client"
	],
	scripts: { build: "vite build && tsc" },
	engines: { node: "^20.19.0 || >=22.12.0" },
	type: "module",
	exports: { ".": "./src/index.ts" },
	publishConfig: {
		access: "public",
		exports: {
			".": "./dist/index.js",
			"./package.json": "./package.json"
		}
	},
	sideEffects: !1,
	files: [
		"dist",
		"src",
		"templates"
	],
	peerDependencies: { meilisearch: "catalog:" }
}, K = (e = []) => {
	let t = `Meilisearch instant-meilisearch (v${G.version})`;
	return e.concat(t);
};
//#endregion
//#region src/client/instant-meilisearch-client.ts
function q(t, o = "", s = {}) {
	i(t, o, s), o = r(o);
	let c = K(s.clientAgents), l = {
		host: t,
		apiKey: o,
		clientAgents: c
	};
	s.httpClient !== void 0 && (l.httpClient = s.httpClient), s.requestInit !== void 0 && (l.requestInit = s.requestInit);
	let u = new e(l), d = V(), f = a(u, d), p = {}, m = n(s);
	return {
		meiliSearchInstance: u,
		setMeiliSearchParams: (e) => {
			let { meiliSearchParams: t } = s;
			s.meiliSearchParams = t === void 0 ? e : {
				...t,
				...e,
				indexesOverrides: {
					...t.indexesOverrides || {},
					...e.indexesOverrides || {}
				}
			};
		},
		searchClient: {
			clearCache: () => d.clearCache(),
			search: async function(e) {
				try {
					let t = [], n = [], r = [];
					for (let i of e) {
						let e = z(i, s), a = y(e);
						t.push(a);
						let o = H(e);
						r.push(o), n.push(e.pagination);
					}
					p = await U(f, r, p);
					let i = await f.multiSearch(t, n);
					return p = W(p, i), P(i, p, m);
				} catch (e) {
					throw console.error(e), Error(e?.message ?? String(e), { cause: e });
				}
			},
			searchForFacetValues: async function(e) {
				let t = [];
				for (let n of e) {
					let e = B(n, s), r = {
						...y(e),
						facetQuery: n.params.facetQuery,
						facetName: n.params.facetName
					};
					delete r.indexUid;
					let i = await u.index(e.indexUid).searchForFacetValues(r), a = {
						facetHits: i.facetHits.map((e) => ({
							...e,
							highlighted: e.value
						})),
						exhaustiveFacetsCount: !1,
						processingTimeMS: i.processingTimeMs
					};
					t.push(a);
				}
				return t;
			}
		}
	};
}
//#endregion
//#region src/analytics/analytics-metadata.ts
function J(e, t) {
	return e?._rawResults ? t?.indexUid ? e._rawResults.find((e) => e.index === t.indexUid)?._meilisearch?.metadata : e._rawResults.map((e) => e._meilisearch?.metadata).filter((e) => !!e) : t ? void 0 : [];
}
//#endregion
export { J as getAnalyticsMetadata, q as instantMeiliSearch };
