// instantsearch-ui-components/dist/cjs/lib/utils/chat.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: Object.getOwnPropertyDescriptor(all, name).get
    });
}
_export(exports, {
    get findTool () {
        return findTool;
    },
    get getFacetFiltersFromToolInput () {
        return getFacetFiltersFromToolInput;
    },
    get getHitsByObjectID () {
        return getHitsByObjectID;
    },
    get getTextContent () {
        return getTextContent;
    },
    get hasTextContent () {
        return hasTextContent;
    },
    get isPartText () {
        return isPartText;
    },
    get isPartTool () {
        return isPartTool;
    }
});
var _sliced_to_array = require("@swc/helpers/_/_sliced_to_array");
var _startsWith = require("./startsWith");
// Keep in sync with packages/instantsearch.js/src/lib/chat/index.ts
var SearchIndexToolType = 'algolia_search_index';
var getTextContent = function getTextContent(message) {
    return message.parts.map(function(part) {
        return 'text' in part ? part.text : '';
    }).join('');
};
var hasTextContent = function hasTextContent(message) {
    return getTextContent(message).trim() !== '';
};
var isPartText = function isPartText(part) {
    return part.type === 'text';
};
var isPartTool = function isPartTool(part) {
    return (0, _startsWith.startsWith)(part.type, 'tool-');
};
var findTool = function findTool(partType, tools) {
    var toolName = partType.replace('tool-', '');
    var tool = tools[toolName];
    if (!tool) {
        var _Object_entries_find;
        tool = (_Object_entries_find = Object.entries(tools).find(function(param) {
            var _param = _sliced_to_array._(param, 1), key = _param[0];
            return (0, _startsWith.startsWith)(toolName, "".concat(key, "_"));
        })) === null || _Object_entries_find === void 0 ? void 0 : _Object_entries_find[1];
    }
    return tool;
};
var FACET_KEY_PREFIX = 'facet_';
var getFacetFiltersFromToolInput = function getFacetFiltersFromToolInput(input) {
    if (!input) {
        return undefined;
    }
    if (Array.isArray(input.facet_filters)) {
        return input.facet_filters;
    }
    var facetFilters = Object.entries(input).reduce(function(acc, param) {
        var _param = _sliced_to_array._(param, 2), key = _param[0], value = _param[1];
        if (!(0, _startsWith.startsWith)(key, FACET_KEY_PREFIX) || !Array.isArray(value)) {
            return acc;
        }
        var attribute = key.slice(FACET_KEY_PREFIX.length);
        var values = value.filter(function(item) {
            return typeof item === 'string';
        });
        if (attribute && values.length > 0) {
            acc.push(values.map(function(item) {
                return "".concat(attribute, ":").concat(item);
            }));
        }
        return acc;
    }, []);
    return facetFilters.length > 0 ? facetFilters : undefined;
};
var isSearchToolPart = function isSearchToolPart(part) {
    return part.type === "tool-".concat(SearchIndexToolType) || // Compatibility shim with Algolia MCP Server search tool
    (0, _startsWith.startsWith)(part.type, "tool-".concat(SearchIndexToolType, "_"));
};
var collectHitsFromPart = function collectHitsFromPart(part, hitsByObjectID) {
    var output = part.state === 'output-available' ? part.output : undefined;
    var hits = output === null || output === void 0 ? void 0 : output.hits;
    if (!Array.isArray(hits)) {
        return;
    }
    hits.forEach(function(hit) {
        if (!hit) {
            return;
        }
        if (typeof hit.objectID === 'string' && hit.objectID !== '') {
            hitsByObjectID[hit.objectID] = hit;
        }
    });
};
var getHitsByObjectID = function getHitsByObjectID(messages, untilToolCallId) {
    var hitsByObjectID = {};
    var _iteratorNormalCompletion = true, _didIteratorError = false, _iteratorError = undefined;
    try {
        var _loop = function() {
            var message = _step.value;
            var reachedBoundary = false;
            message.parts.forEach(function(part) {
                if (!isPartTool(part)) {
                    return;
                }
                // Note the boundary but keep processing the rest of this message's parts:
                // the search tool that fed this display tool lives in the same message,
                // so its hits must still be collected before we stop.
                if (untilToolCallId && part.toolCallId === untilToolCallId) {
                    reachedBoundary = true;
                }
                if (isSearchToolPart(part)) {
                    collectHitsFromPart(part, hitsByObjectID);
                }
            });
            if (reachedBoundary) {
                return "break";
            }
        };
        for(var _iterator = messages[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true){
            var _ret = _loop();
            if (_ret === "break") break;
        }
    } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
    } finally{
        try {
            if (!_iteratorNormalCompletion && _iterator.return != null) {
                _iterator.return();
            }
        } finally{
            if (_didIteratorError) {
                throw _iteratorError;
            }
        }
    }
    return hitsByObjectID;
};
