// instantsearch.js/es/connectors/feeds/connectFeeds.js
import { _ as _$2 } from '@swc/helpers/esm/_instanceof.js';
import { _ as _$1 } from '@swc/helpers/esm/_object_spread.js';
import { _ } from '@swc/helpers/esm/_object_spread_props.js';
import algoliasearchHelper from 'algoliasearch-helper';
import { createDocumentationMessageGenerator } from '../../lib/utils/documentation.js';
import { checkRendering } from '../../lib/utils/checkRendering.js';
import { noop } from '../../lib/utils/noop.js';

function toFeedSearchResults(state, raw) {
    return Object.assign(new algoliasearchHelper.SearchResults(state, [
        raw
    ]), {
        feedID: raw.feedID
    });
}
/**
 * Rebuild `lastResults.feeds` from `_initialResults.compositionFeedsResults`
 * because the index-widget hydration only restores `lastResults` (the merged
 * view), not the per-feed breakdown that the Feeds connector needs.
 */ function hydrateFeedsFromInitialResultsIfNeeded(instantSearchInstance, parent) {
    var _instantSearchInstance__initialResults, _parent_getHelper;
    var initial = (_instantSearchInstance__initialResults = instantSearchInstance._initialResults) === null || _instantSearchInstance__initialResults === void 0 ? void 0 : _instantSearchInstance__initialResults[parent.getIndexId()];
    var compositionFeedsResults = (initial === null || initial === void 0 ? void 0 : initial.compositionFeedsResults) || [];
    if (compositionFeedsResults.length === 0) {
        return;
    }
    var lastResults = (_parent_getHelper = parent.getHelper()) === null || _parent_getHelper === void 0 ? void 0 : _parent_getHelper.lastResults;
    if (!lastResults) {
        return;
    }
    if (lastResults.feeds && lastResults.feeds.length > 0) {
        return;
    }
    lastResults.feeds = compositionFeedsResults.map(function(raw) {
        return toFeedSearchResults(lastResults._state, raw);
    });
}
var withUsage = createDocumentationMessageGenerator({
    name: 'feeds',
    connector: true
});
var connectFeeds = function connectFeeds(renderFn) {
    var unmountFn = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : noop;
    checkRendering(renderFn, withUsage());
    return function(widgetParams) {
        var isolated = widgetParams.isolated, _widgetParams_transformFeeds = widgetParams.transformFeeds, transformFeeds = _widgetParams_transformFeeds === void 0 ? function(feeds) {
            return feeds;
        } : _widgetParams_transformFeeds;
        if (isolated !== false) {
            throw new Error(withUsage('The `isolated` option currently only supports `false`.'));
        }
        return {
            $$type: 'ais.feeds',
            $$widgetType: 'ais.feeds',
            init: function init(initOptions) {
                var instantSearchInstance = initOptions.instantSearchInstance;
                if (!instantSearchInstance.compositionID) {
                    throw new Error(withUsage('The `feeds` widget requires a composition-based InstantSearch instance (compositionID must be set).'));
                }
                hydrateFeedsFromInitialResultsIfNeeded(instantSearchInstance, initOptions.parent);
                renderFn(_(_$1({}, this.getWidgetRenderState(initOptions)), {
                    instantSearchInstance: instantSearchInstance
                }), true);
            },
            render: function render(renderOptions) {
                var instantSearchInstance = renderOptions.instantSearchInstance;
                renderFn(_(_$1({}, this.getWidgetRenderState(renderOptions)), {
                    instantSearchInstance: instantSearchInstance
                }), false);
            },
            dispose: function dispose() {
                unmountFn();
            },
            getWidgetSearchParameters: function getWidgetSearchParameters(state) {
                return state;
            },
            getRenderState: function getRenderState(renderState, renderOptions) {
                return _(_$1({}, renderState), {
                    feeds: this.getWidgetRenderState(renderOptions)
                });
            },
            getWidgetRenderState: function getWidgetRenderState(param) {
                var results = param.results;
                if (!results) {
                    return {
                        feedIDs: [],
                        widgetParams: widgetParams
                    };
                }
                if (Array.isArray(results.feeds) && results.feeds.length > 0 && !results.feeds.every(function(feed) {
                    return _$2(feed, algoliasearchHelper.SearchResults);
                })) {
                    results.feeds = results.feeds.map(function(feed) {
                        return _$2(feed, algoliasearchHelper.SearchResults) ? feed : toFeedSearchResults(results._state, feed);
                    });
                }
                var feedIDs = results.feeds ? results.feeds.map(function(f) {
                    return f.feedID;
                }) : [
                    ''
                ];
                feedIDs = transformFeeds(feedIDs);
                if (!Array.isArray(feedIDs)) {
                    throw new Error(withUsage('The `transformFeeds` option expects a function that returns an Array.'));
                }
                if (!feedIDs.every(function(feedID) {
                    return typeof feedID === 'string';
                })) {
                    throw new Error(withUsage('The `transformFeeds` option expects a function that returns an array of feed IDs (strings).'));
                }
                return {
                    feedIDs: feedIDs,
                    widgetParams: widgetParams
                };
            }
        };
    };
};

export { connectFeeds as default };
