// instantsearch.js/es/connectors/feeds/FeedContainer.js
import { _ } from '@swc/helpers/esm/_instanceof.js';
import algoliasearchHelper from 'algoliasearch-helper';
import { createInitArgs, storeRenderState, createRenderArgs } from '../../lib/utils/render-args.js';

function createFeedContainer(feedID, parentIndex, instantSearchInstance) {
    var localWidgets = [];
    var initialized = false;
    var container = {
        $$type: 'ais.feedContainer',
        $$widgetType: 'ais.feedContainer',
        _isolated: true,
        getIndexName: function getIndexName() {
            return parentIndex.getIndexName();
        },
        getIndexId: function getIndexId() {
            return feedID;
        },
        getHelper: function getHelper() {
            return parentIndex.getHelper();
        },
        getResults: function getResults() {
            var parentResults = parentIndex.getResults();
            if (!parentResults) return null;
            if (!parentResults.feeds) {
                // Single-feed backward compat: no feeds array means the parent result
                // itself is the only feed.
                if (feedID === '') {
                    parentResults._state = parentIndex.getHelper().state;
                    return parentResults;
                }
                return null;
            }
            var feed = parentResults.feeds.find(function(f) {
                return f.feedID === feedID;
            });
            if (!feed) return null;
            // Optimistic state patching — same as index widget (index.ts:365-370)
            feed._state = parentIndex.getHelper().state;
            return feed;
        },
        getResultsForWidget: function getResultsForWidget() {
            return this.getResults();
        },
        getParent: function getParent() {
            return parentIndex;
        },
        getWidgets: function getWidgets() {
            return localWidgets;
        },
        getScopedResults: function getScopedResults() {
            return parentIndex.getScopedResults();
        },
        getPreviousState: function getPreviousState() {
            return null;
        },
        createURL: function createURL(nextState) {
            return parentIndex.createURL(nextState);
        },
        scheduleLocalSearch: function scheduleLocalSearch() {
            return parentIndex.scheduleLocalSearch();
        },
        addWidgets: function addWidgets(widgets) {
            var flatWidgets = widgets.reduce(function(acc, w) {
                return acc.concat(Array.isArray(w) ? w : [
                    w
                ]);
            }, []);
            flatWidgets.forEach(function(widget) {
                widget.parent = container;
            });
            localWidgets = localWidgets.concat(flatWidgets);
            if (initialized) {
                flatWidgets.forEach(function(widget) {
                    if (widget.getRenderState) {
                        var renderState = widget.getRenderState(instantSearchInstance.renderState[container.getIndexId()] || {}, createInitArgs(instantSearchInstance, container, instantSearchInstance._initialUiState));
                        storeRenderState({
                            renderState: renderState,
                            instantSearchInstance: instantSearchInstance,
                            parent: container
                        });
                    }
                });
                flatWidgets.forEach(function(widget) {
                    if (widget.init) {
                        widget.init(createInitArgs(instantSearchInstance, container, instantSearchInstance._initialUiState));
                    }
                });
                // Merge children's search params (e.g. disjunctiveFacets) into the
                // parent's helper state so they're included in the composition request.
                // uiState is {} because URL-derived refinements are already on the
                // parent state; children only need to declare structural params.
                var parentHelper = parentIndex.getHelper();
                var withChildParams = container.getWidgetSearchParameters(parentHelper.state, {
                    uiState: {}
                });
                if (withChildParams !== parentHelper.state) {
                    parentHelper.setState(withChildParams);
                }
            }
            return container;
        },
        removeWidgets: function removeWidgets(widgets) {
            var flatWidgets = widgets.reduce(function(acc, w) {
                return acc.concat(Array.isArray(w) ? w : [
                    w
                ]);
            }, []);
            var helper = parentIndex.getHelper();
            if (!helper) {
                localWidgets = localWidgets.filter(function(w) {
                    return !flatWidgets.includes(w);
                });
                return container;
            }
            // Chain through children's dispose so widgets clean up the
            // SearchParameters they declared (e.g. RefinementList removes its
            // disjunctiveFacet) instead of leaving them stale on the parent helper.
            var cleanedState = helper.state;
            flatWidgets.forEach(function(widget) {
                if (widget.dispose) {
                    var next = widget.dispose({
                        helper: helper,
                        state: cleanedState,
                        recommendState: helper.recommendState,
                        parent: container
                    });
                    if (_(next, algoliasearchHelper.RecommendParameters)) ; else if (next) {
                        cleanedState = next;
                    }
                }
            });
            localWidgets = localWidgets.filter(function(w) {
                return !flatWidgets.includes(w);
            });
            if (cleanedState !== helper.state) {
                helper.setState(cleanedState);
            }
            return container;
        },
        init: function init() {
            initialized = true;
            localWidgets.forEach(function(widget) {
                if (widget.getRenderState) {
                    var renderState = widget.getRenderState(instantSearchInstance.renderState[container.getIndexId()] || {}, createInitArgs(instantSearchInstance, container, instantSearchInstance._initialUiState));
                    storeRenderState({
                        renderState: renderState,
                        instantSearchInstance: instantSearchInstance,
                        parent: container
                    });
                }
            });
            localWidgets.forEach(function(widget) {
                if (widget.init) {
                    widget.init(createInitArgs(instantSearchInstance, container, instantSearchInstance._initialUiState));
                }
            });
        },
        render: function render() {
            localWidgets.forEach(function(widget) {
                if (widget.getRenderState) {
                    var renderState = widget.getRenderState(instantSearchInstance.renderState[container.getIndexId()] || {}, createRenderArgs(instantSearchInstance, container, widget));
                    storeRenderState({
                        renderState: renderState,
                        instantSearchInstance: instantSearchInstance,
                        parent: container
                    });
                }
            });
            localWidgets.forEach(function(widget) {
                if (widget.render) {
                    widget.render(createRenderArgs(instantSearchInstance, container, widget));
                }
            });
        },
        dispose: function dispose(disposeOptions) {
            var _ref;
            var helper = parentIndex.getHelper();
            // Chain through children's dispose to return a cleaned state
            // (e.g. RefinementList.dispose removes its disjunctiveFacet declaration).
            // This mirrors how the index widget's removeWidgets chains dispose calls.
            var cleanedState = (_ref = disposeOptions === null || disposeOptions === void 0 ? void 0 : disposeOptions.state) !== null && _ref !== void 0 ? _ref : helper === null || helper === void 0 ? void 0 : helper.state;
            localWidgets.forEach(function(widget) {
                if (widget.dispose && helper) {
                    var next = widget.dispose({
                        helper: helper,
                        state: cleanedState,
                        recommendState: helper.recommendState,
                        parent: container
                    });
                    if (_(next, algoliasearchHelper.RecommendParameters)) ; else if (next) {
                        cleanedState = next;
                    }
                }
            });
            localWidgets = [];
            initialized = false;
            return cleanedState;
        },
        getWidgetState: function getWidgetState(uiState) {
            return this.getWidgetUiState(uiState);
        },
        getWidgetUiState: function getWidgetUiState(uiState) {
            var helper = parentIndex.getHelper();
            var widgetUiStateOptions = {
                searchParameters: helper.state,
                helper: helper
            };
            return localWidgets.reduce(function(state, widget) {
                return widget.getWidgetUiState ? widget.getWidgetUiState(state, widgetUiStateOptions) : state;
            }, uiState);
        },
        getWidgetSearchParameters: function getWidgetSearchParameters(searchParameters, param) {
            var uiState = param.uiState;
            return localWidgets.reduce(function(params, widget) {
                return widget.getWidgetSearchParameters ? widget.getWidgetSearchParameters(params, {
                    uiState: uiState
                }) : params;
            }, searchParameters);
        },
        refreshUiState: function refreshUiState() {
        // no-op: FeedContainer doesn't own UI state
        },
        setIndexUiState: function setIndexUiState() {
        // no-op: FeedContainer delegates to parent
        }
    };
    return container;
}

export { createFeedContainer };
