// instantsearch.js/es/widgets/feeds/feeds.js
import { _ as _$3 } from '@swc/helpers/esm/_object_spread.js';
import { _ as _$2 } from '@swc/helpers/esm/_object_spread_props.js';
import { _ } from '@swc/helpers/esm/_object_without_properties.js';
import { _ as _$1 } from '@swc/helpers/esm/_sliced_to_array.js';
import { _ as _$4 } from '@swc/helpers/esm/_to_consumable_array.js';
import connectFeeds from '../../connectors/feeds/connectFeeds.js';
import { createFeedContainer } from '../../connectors/feeds/FeedContainer.js';
import { component } from '../../lib/suit.js';
import { createDocumentationMessageGenerator } from '../../lib/utils/documentation.js';
import { getContainerNode } from '../../lib/utils/getContainerNode.js';

var withUsage = createDocumentationMessageGenerator({
    name: 'feeds'
});
var suit = component('Feeds');
var feeds = function feeds(widgetParams) {
    var _ref = widgetParams || {}, containerSelector = _ref.container, widgets = _ref.widgets, transformFeeds = _ref.transformFeeds, connectorParams = _(_ref, [
        "container",
        "widgets",
        "transformFeeds"
    ]);
    if (!containerSelector) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    if (typeof widgets !== 'function') {
        throw new Error(withUsage('The `widgets` option expects a function.'));
    }
    var userContainer = getContainerNode(containerSelector);
    var rootContainer = document.createElement('div');
    rootContainer.className = suit();
    var feedEntries = new Map();
    var pendingRemovals = new Map();
    var removalTimer = null;
    var parentRef;
    var renderOptionsRef;
    var makeWidget = connectFeeds(function(param, isFirstRender) {
        var feedIDs = param.feedIDs;
        if (isFirstRender) {
            userContainer.appendChild(rootContainer);
            return;
        }
        var activeFeedIDs = new Set(feedIDs);
        // Remove containers for feeds that no longer exist (deferred)
        var toRemove = [];
        feedEntries.forEach(function(entry, id) {
            if (!activeFeedIDs.has(id)) {
                toRemove.push([
                    id,
                    entry
                ]);
                rootContainer.removeChild(entry.domElement);
                feedEntries.delete(id);
            }
        });
        if (toRemove.length > 0) {
            toRemove.forEach(function(param) {
                var _param = _$1(param, 2), id = _param[0], entry = _param[1];
                pendingRemovals.set(id, entry);
            });
            if (removalTimer !== null) {
                clearTimeout(removalTimer);
            }
            removalTimer = setTimeout(function() {
                var widgetsToRemove = Array.from(pendingRemovals.values()).map(function(entry) {
                    return entry.feedContainer;
                });
                pendingRemovals.clear();
                removalTimer = null;
                if (widgetsToRemove.length > 0) {
                    parentRef.removeWidgets(widgetsToRemove);
                }
            }, 0);
        }
        // Create containers for new feeds
        feedIDs.forEach(function(feedID) {
            if (!feedEntries.has(feedID)) {
                var pendingEntry = pendingRemovals.get(feedID);
                if (pendingEntry) {
                    pendingRemovals.delete(feedID);
                    feedEntries.set(feedID, pendingEntry);
                    rootContainer.appendChild(pendingEntry.domElement);
                    pendingEntry.feedContainer.render(renderOptionsRef);
                    return;
                }
                var domElement = document.createElement('div');
                domElement.className = suit({
                    descendantName: 'feed'
                });
                var feedWidgets = widgets(domElement, feedID);
                var feedContainer = createFeedContainer(feedID, parentRef, renderOptionsRef.instantSearchInstance);
                feedEntries.set(feedID, {
                    feedContainer: feedContainer,
                    domElement: domElement
                });
                rootContainer.appendChild(domElement);
                parentRef.addWidgets([
                    feedContainer
                ]);
                if (feedWidgets && feedWidgets.length > 0) {
                    feedContainer.addWidgets(feedWidgets);
                }
                feedContainer.render(renderOptionsRef);
            }
        });
        // Reorder DOM nodes to match feedIDs order
        feedIDs.forEach(function(feedID) {
            var entry = feedEntries.get(feedID);
            if (entry) {
                rootContainer.appendChild(entry.domElement);
            }
        });
    }, function() {
        userContainer.removeChild(rootContainer);
    });
    var widget = makeWidget(_$2(_$3({}, connectorParams), {
        transformFeeds: transformFeeds
    }));
    return _$2(_$3({}, widget), {
        init: function init(initOptions) {
            parentRef = initOptions.parent;
            widget.init(initOptions);
        },
        render: function render(renderOptions) {
            parentRef = renderOptions.parent;
            renderOptionsRef = renderOptions;
            widget.render(renderOptions);
        },
        dispose: function dispose(disposeOptions) {
            if (removalTimer !== null) {
                clearTimeout(removalTimer);
                removalTimer = null;
            }
            var activeContainers = Array.from(feedEntries.values()).map(function(entry) {
                return entry.feedContainer;
            });
            var pendingContainers = Array.from(pendingRemovals.values()).map(function(entry) {
                return entry.feedContainer;
            });
            var toRemove = Array.from(new Set(_$4(activeContainers).concat(_$4(pendingContainers))));
            pendingRemovals.clear();
            feedEntries.clear();
            if (toRemove.length > 0) {
                disposeOptions.parent.removeWidgets(toRemove);
            }
            widget.dispose(disposeOptions);
        },
        $$widgetType: 'ais.feeds'
    });
};

export { feeds as default };
