// instantsearch.js/cjs/widgets/feeds/feeds.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
var _interop_require_default = require("@swc/helpers/_/_interop_require_default");
var _object_spread = require("@swc/helpers/_/_object_spread");
var _object_spread_props = require("@swc/helpers/_/_object_spread_props");
var _object_without_properties = require("@swc/helpers/_/_object_without_properties");
var _sliced_to_array = require("@swc/helpers/_/_sliced_to_array");
var _to_consumable_array = require("@swc/helpers/_/_to_consumable_array");
var _connectFeeds = /*#__PURE__*/ _interop_require_default._(require("../../connectors/feeds/connectFeeds"));
var _FeedContainer = require("../../connectors/feeds/FeedContainer");
var _suit = require("../../lib/suit");
var _utils = require("../../lib/utils");
var withUsage = (0, _utils.createDocumentationMessageGenerator)({
    name: 'feeds'
});
var suit = (0, _suit.component)('Feeds');
var feeds = function feeds(widgetParams) {
    var _ref = widgetParams || {}, containerSelector = _ref.container, widgets = _ref.widgets, transformFeeds = _ref.transformFeeds, connectorParams = _object_without_properties._(_ref, [
        "container",
        "widgets",
        "transformFeeds"
    ]);
    if (!containerSelector) {
        throw new Error(withUsage('The `container` option is required.'));
    }
    if (typeof widgets !== 'function') {
        throw new Error(withUsage('The `widgets` option expects a function.'));
    }
    var userContainer = (0, _utils.getContainerNode)(containerSelector);
    var rootContainer = document.createElement('div');
    rootContainer.className = suit();
    var feedEntries = new Map();
    var pendingRemovals = new Map();
    var removalTimer = null;
    var parentRef;
    var renderOptionsRef;
    var makeWidget = (0, _connectFeeds.default)(function(param, isFirstRender) {
        var feedIDs = param.feedIDs;
        if (isFirstRender) {
            userContainer.appendChild(rootContainer);
            return;
        }
        var activeFeedIDs = new Set(feedIDs);
        // Remove containers for feeds that no longer exist (deferred)
        var toRemove = [];
        feedEntries.forEach(function(entry, id) {
            if (!activeFeedIDs.has(id)) {
                toRemove.push([
                    id,
                    entry
                ]);
                rootContainer.removeChild(entry.domElement);
                feedEntries.delete(id);
            }
        });
        if (toRemove.length > 0) {
            toRemove.forEach(function(param) {
                var _param = _sliced_to_array._(param, 2), id = _param[0], entry = _param[1];
                pendingRemovals.set(id, entry);
            });
            if (removalTimer !== null) {
                clearTimeout(removalTimer);
            }
            removalTimer = setTimeout(function() {
                var widgetsToRemove = Array.from(pendingRemovals.values()).map(function(entry) {
                    return entry.feedContainer;
                });
                pendingRemovals.clear();
                removalTimer = null;
                if (widgetsToRemove.length > 0) {
                    parentRef.removeWidgets(widgetsToRemove);
                }
            }, 0);
        }
        // Create containers for new feeds
        feedIDs.forEach(function(feedID) {
            if (!feedEntries.has(feedID)) {
                var pendingEntry = pendingRemovals.get(feedID);
                if (pendingEntry) {
                    pendingRemovals.delete(feedID);
                    feedEntries.set(feedID, pendingEntry);
                    rootContainer.appendChild(pendingEntry.domElement);
                    pendingEntry.feedContainer.render(renderOptionsRef);
                    return;
                }
                var domElement = document.createElement('div');
                domElement.className = suit({
                    descendantName: 'feed'
                });
                var feedWidgets = widgets(domElement, feedID);
                var feedContainer = (0, _FeedContainer.createFeedContainer)(feedID, parentRef, renderOptionsRef.instantSearchInstance);
                feedEntries.set(feedID, {
                    feedContainer: feedContainer,
                    domElement: domElement
                });
                rootContainer.appendChild(domElement);
                parentRef.addWidgets([
                    feedContainer
                ]);
                if (feedWidgets && feedWidgets.length > 0) {
                    feedContainer.addWidgets(feedWidgets);
                }
                feedContainer.render(renderOptionsRef);
            }
        });
        // Reorder DOM nodes to match feedIDs order
        feedIDs.forEach(function(feedID) {
            var entry = feedEntries.get(feedID);
            if (entry) {
                rootContainer.appendChild(entry.domElement);
            }
        });
    }, function() {
        userContainer.removeChild(rootContainer);
    });
    var widget = makeWidget(_object_spread_props._(_object_spread._({}, connectorParams), {
        transformFeeds: transformFeeds
    }));
    return _object_spread_props._(_object_spread._({}, widget), {
        init: function init(initOptions) {
            parentRef = initOptions.parent;
            widget.init(initOptions);
        },
        render: function render(renderOptions) {
            parentRef = renderOptions.parent;
            renderOptionsRef = renderOptions;
            widget.render(renderOptions);
        },
        dispose: function dispose(disposeOptions) {
            if (removalTimer !== null) {
                clearTimeout(removalTimer);
                removalTimer = null;
            }
            var activeContainers = Array.from(feedEntries.values()).map(function(entry) {
                return entry.feedContainer;
            });
            var pendingContainers = Array.from(pendingRemovals.values()).map(function(entry) {
                return entry.feedContainer;
            });
            var toRemove = Array.from(new Set(_to_consumable_array._(activeContainers).concat(_to_consumable_array._(pendingContainers))));
            pendingRemovals.clear();
            feedEntries.clear();
            if (toRemove.length > 0) {
                disposeOptions.parent.removeWidgets(toRemove);
            }
            widget.dispose(disposeOptions);
        },
        $$widgetType: 'ais.feeds'
    });
};
var _default = feeds;
